packages/compiler/core/checker.ts

```diff
@@ -2167,18 +2315,27 @@ export function createChecker(program: Program): Checker {
     }
   }
 
-  function checkUnionVariant(variantNode: UnionVariantNode): UnionTypeVariant {
+  function checkUnionVariant(
+    union: UnionStatementNode,
+    variantNode: UnionVariantNode
+  ): UnionTypeVariant {
     const name =
       variantNode.id.kind === SyntaxKind.Identifier ? variantNode.id.sv : variantNode.id.value;
     const decorators = checkDecorators(variantNode);
     const type = getTypeForNode(variantNode.value);
-    return createAndFinishType({
+    const variantType: UnionTypeVariant = createType({
       kind: "UnionVariant",
       name,
       node: variantNode,
       decorators,
       type,
     });
+
+    if (shouldCreateTypeForTemplate(union)) {
```

Comment:
Fixed this bug I ran in to: we were running decorators on uninstantiated union variants.